sql/src/main/java/io/crate/metadata/Functions.java

```diff
@@ -31,73 +31,123 @@
+        schemaFunctionResolvers = new ConcurrentHashMap<>();
     }
 
-    private void generateFunctionResolvers(Map<FunctionIdent, FunctionImplementation> functionImplementations) {
+    public Map<String, FunctionResolver> generateFunctionResolvers(Map<FunctionIdent, FunctionImplementation> functionImplementations) {
+        Multimap<String, Tuple<FunctionIdent, FunctionImplementation>> signatures = getSignatures(functionImplementations);
+        return signatures.keys().stream()
+            .distinct()
+            .collect(Collectors.toMap(name -> name, name -> new GeneratedFunctionResolver(signatures.get(name))));
+    }
+
+    private Multimap<String, Tuple<FunctionIdent, FunctionImplementation>> getSignatures(
+        Map<FunctionIdent, FunctionImplementation> functionImplementations) {
         Multimap<String, Tuple<FunctionIdent, FunctionImplementation>> signatureMap = ArrayListMultimap.create();
         for (Map.Entry<FunctionIdent, FunctionImplementation> entry : functionImplementations.entrySet()) {
             signatureMap.put(entry.getKey().name(), new Tuple<>(entry.getKey(), entry.getValue()));
         }
-        for (String name : signatureMap.keys()) {
-            functionResolvers.put(name, new GeneratedFunctionResolver(signatureMap.get(name)));
+        return signatureMap;
+    }
+
+    public void registerSchemaFunctionResolvers(String schema, Map<String, FunctionResolver> resolvers) {
+        schemaFunctionResolvers.put(schema, resolvers);
+    }
+
+    /**
+     * Returns the function implementation for the given schema, function name and arguments.
+     *
+     * If the explicit schema name is not provided then the first function look-up will be done
+     * in the built-in functions. If the function is not found there, then the method will
+     * check the default schema for it.
+     *
+     * In case when the schema name is given, the look-up is done directly for the schema name.
+     *
+     * @param schema    The schema name. The schema name is null for built-in functions.
+     * @param name      The function name.
+     * @param arguments The function arguments.
+     * @return a function implementation or null if nothing was found.
+     */
+    @Nullable
+    public FunctionImplementation get(@Nullable String schema, String name, List<DataType> arguments)
+        throws IllegalArgumentException {
+        // get a built-in function
+        if (schema == null) {
+            FunctionResolver dynamicResolver = functionResolvers.get(name);
+            if (dynamicResolver != null) {
+                List<DataType> signature = dynamicResolver.getSignature(arguments);
+                if (signature != null) {
+                    return dynamicResolver.getForTypes(signature);
+                }
+            }
+            // check if the function without explicit schema name exists in the default schema
+            return getSchemaFunctionImplementation(Schemas.DEFAULT_SCHEMA_NAME, name, arguments);
```

Comment:
Doesn't this have to fallback to sessionContext.defaultSchema (and only if that is false, use DOC)